crates/autopilot/src/run_loop.rs

```diff
@@ -389,6 +399,83 @@ impl RunLoop {
         .collect()
     }
 
+    /// Returns true if winning solution is fair or winner is None
+    fn check_fairness(
+        winner: Option<&Participant>,
+        remaining: &Vec<Participant>,
+        auction: &domain::Auction,
+    ) -> bool {
+        let winner = match winner {
+            Some(winner) => winner,
+            None => return true,
+        };
+        let fairness_threshold = match winner.driver.fairness_threshold {
+            Some(threshold) => threshold,
+            None => return true,
+        };
+
+        // Returns Some(buy token amount) if left is better than right.
+        let improvement_in_buy = |left: &TradedAmounts, right: &TradedAmounts| {
+            right
+                .sell
+                .checked_mul(&left.buy)
+                .unwrap_or(U256::max_value().into())
+                .checked_sub(
+                    &left
+                        .sell
+                        .checked_mul(&right.buy)
+                        .unwrap_or(U256::max_value().into()),
+                )?
+                .checked_div(&right.sell)
+                .filter(|diff| !diff.is_zero())
+        };
+
+        // Find best execution per order
+        let mut best_executions = HashMap::new();
+        for other in remaining {
+            for (uid, execution) in other.solution.orders() {
+                let best_execution = best_executions.entry(uid).or_insert(execution);
+                if improvement_in_buy(execution, best_execution).is_some() {
+                    *best_execution = execution;
+                }
```

Comment:
I honestly think it's ok to have a best_execution contain the execution of the winning solution if that's the best execution of the order (note that `improvement` returns `None` in case of equality)

Reply:
Yeah, I didn't mean to not put it into the map. Just that we always put the first execution of an order into the map and then afterwards compare it against itself. But given that most of my remarks are about readability of this code and this would not improve that aspect I'll drop my remark.